### ISSUE TITLE
JIT failed on Jupyter Notebook file.

### ISSUE THREAD

**Issue by Honghe** [cupy/cupy#6320]
### Description

Jupyter Notebook output:
```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Input In [1], in <module>
      3 a = cupy.array([0.4, -0.2, 1.8, -1.2])
      4 relu = cupy.vectorize(lambda x: (x > 0.0) * x)
----> 5 print(relu(a))

File ~\miniconda3\envs\cuda\lib\site-packages\cupy\_functional\vectorize.py:87, in vectorize.__call__(self, *args)
     84     raise NotImplementedError
     86 func = _interface._CudaFunction(self.pyfunc, 'numpy', device=True)
---> 87 result = func._emit_code_from_types(in_types, ret_type)
     88 in_params = ', '.join(
     89     f'{t.dtype} in{i}' for i, t in enumerate(in_types))
     90 in_args = ', '.join([f'in{i}' for i in range(len(in_types))])

File ~\miniconda3\envs\cuda\lib\site-packages\cupyx\jit\_interface.py:38, in _CudaFunction._emit_code_from_types(self, in_types, ret_type)
     37 def _emit_code_from_types(self, in_types, ret_type=None):
---> 38     return _compile.transpile(
     39         self.func, self.attributes, self.mode, in_types, ret_type)

File ~\miniconda3\envs\cuda\lib\site-packages\cupyx\jit\_compile.py:123, in transpile(func, attributes, mode, in_types, ret_type)
    121 consts = dict(**cvars.globals, **cvars.nonlocals, **cvars.builtins)
    122 attributes = ' '.join(attributes)
--> 123 tree, source = _parse_function_object(func)
    124 cuda_code, env = _transpile_function(
    125     tree, attributes, mode, consts, in_types, ret_type, source=source)
    126 cuda_code = ''.join([code + '\n' for code in env.preambles]) + cuda_code

File ~\miniconda3\envs\cuda\lib\site-packages\cupyx\jit\_compile.py:88, in _parse_function_object(func)
     85 if filename is None:
     86     raise ValueError(f'JIT needs access to Python source for {func}'
     87                      'but could not be located')
---> 88 with open(filename) as f:
     89     full_source = f.read()
     90 source, start_line = inspect.getsourcelines(func)

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\honhe\\AppData\\Local\\Temp\\ipykernel_18084\\4273467977.py'
```

### To Reproduce

The source code in Jupyter Notebook file.
```py
import cupy

a = cupy.array([0.4, -0.2, 1.8, -1.2])
relu = cupy.vectorize(lambda x: (x > 0.0) * x)
print(relu(a))  # [ 0.4 -0.   1.8 -0. ]
```

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
> python -c 'import cupy; cupy.show_config()'
C:\Users\honhe\miniconda3\envs\cuda\lib\site-packages\cupy\_environment.py:213: UserWarning: CUDA path could not be detected. Set CUDA_PATH environment variable if CuPy fails to load.
  warnings.warn(
OS                           : Windows-10-10.0.19044-SP0
Python Version               : 3.8.12
CuPy Version                 : 10.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.22.0
SciPy Version                : None
Cython Build Version         : 0.29.25
Cython Runtime Version       : None
CUDA Root                    : None
nvcc PATH                    : None
CUDA Build Version           : 11020
CUDA Driver Version          : 11020
CUDA Runtime Version         : 11020
cuBLAS Version               : (available)
cuFFT Version                : 10401
cuRAND Version               : 10203
cuSOLVER Version             : (11, 1, 0)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 2)
Thrust Version               : 101000
CUB Build Version            : 101000
Jitify Build Version         : <unknown>
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : GeForce MX350
Device 0 Compute Capability  : 61
Device 0 PCI Bus ID          : 0000:01:00.0
```

**Comment by kmaehashi** [comment]
Hi @Honghe, thanks for reporting! Issue confirmed.

We had to use `linecache` here instead of `open` to support `lambda` in notebook environment.
https://github.com/cupy/cupy/pull/6170#discussion_r757727067

**Comment by kmaehashi** [comment]
Will be fixed in #6329
